File: packages/language/src/symbols.ts
import { Module } from "./module.ts";
import { Range } from "./range.ts";
import * as typescript from "./typescript.ts";
import ts from "typescript";

export type Request = {
	module: Module;
};

export type Response = {
	symbols: Array<Symbol> | null;
};

export type Symbol = {
	name: string;
	detail: string | null;
	kind: Kind;
	tags: Array<Tag>;
	range: Range;
	selectionRange: Range;
	children: Array<Symbol> | null;
};

export type Kind =
	| "file"
	| "module"
	| "namespace"
	| "package"
	| "class"
	| "method"
	| "property"
	| "field"
	| "constructor"
	| "enum"
	| "interface"
	| "function"
	| "variable"
	| "constant"
	| "string"
	| "number"
	| "boolean"
	| "array"
	| "object"
	| "key"
	| "null"
	| "enumMember"
	| "event"
	| "operator"
	| "typeParameter";

export type Tag = "deprecated";

export let handle = (request: Request): Response => {
	// Get the module's filename and source.
	let fileName = typescript.fileNameFromModule(request.module);
	let sourceFile = typescript.host.getSourceFile(
		fileName,
		ts.ScriptTarget.ESNext,
	);

	if (sourceFile === undefined) {
		throw new Error();
	}

	// Get the navigation tree for this file.
	let navigationTree = typescript.languageService.getNavigationTree(fileName);

	// Get the symbols by walking the navigation tree.
	let root = walk(sourceFile, navigationTree);

	// Drop the root and return the symbols directly. The item returned by getNavigationTree is guaranteed to be a module.
	return { symbols: root.children };
};

export let walk = (
	sourceFile: ts.SourceFile,
	tree: ts.NavigationTree,
): Symbol => {
	let name = tree.text;

	// Find the range of this symbol and its selectionRange.
	let startPosition = Math.min(...tree.spans.map((span) => span.start));
	let endPosition = Math.max(
		...tree.spans.map((span) => span.start + span.length),
	);

	// Convert text spans to ranges.
	let range = {
		start: ts.getLineAndCharacterOfPosition(sourceFile, startPosition),
		end: ts.getLineAndCharacterOfPosition(sourceFile, endPosition),
	};

	// Parse the symbol kind from the nav tree.
	let kind = getKind(tree.kind);

	// Collect the nested children.
	let children = tree.childItems?.map((child) => walk(sourceFile, child));

	return {
		name,
		kind,
		tags: [],
		detail: null,
		range,
		selectionRange: range,
		children: children ?? null,
	};
};

// Map the "kind" field from NavigationTree to a variant that is meaningful to LSP.
let getKind = (tsKind: string): Kind => {
	switch (tsKind) {
		case "script":
			return "file";
		case "module":
			return "module";
		case "class":
			return "class";
		case "local class":
			return "class";
		case "interface":
			return "interface";
		case "type":
			return "class";
		case "enum":
			return "enum";
		case "var":
			return "variable";
		case "local var":
			return "variable";
		case "function":
			return "function";
		case "local function":
			return "function";
		case "method":
			return "method";
		case "getter":
			return "method";
		case "setter":
			return "method";
		case "property":
			return "property";
		case "accessor":
			return "property";
		case "constructor":
			return "constructor";
		case "parameter":
			return "variable";
		case "type parameter":
			return "typeParameter";
		case "external module":
			return "module";
		default:
			return "variable";
	}
};

File: packages/language/src/symbols.rs
use crate::{Module, Range, Server};
use lsp_types as lsp;
use tangram_error::{error, Result};

#[derive(Debug, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Request {
	pub module: Module,
}

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Response {
	pub symbols: Option<Vec<Symbol>>,
}

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Symbol {
	pub name: String,
	pub detail: Option<String>,
	pub kind: Kind,
	pub tags: Vec<Tag>,
	pub range: Range,
	pub selection_range: Range,
	pub children: Option<Vec<Self>>,
}

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Kind {
	File,
	Module,
	Namespace,
	Package,
	Class,
	Method,
	Property,
	Field,
	Constructor,
	Enum,
	Interface,
	Function,
	Variable,
	Constant,
	String,
	Number,
	Boolean,
	Array,
	Object,
	Key,
	Null,
	EnumMember,
	Event,
	Operator,
	TypeParameter,
}

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Tag {
	Deprecated,
}

impl Server {
	pub(super) async fn handle_symbols_request(
		&self,
		params: lsp::DocumentSymbolParams,
	) -> Result<Option<lsp::DocumentSymbolResponse>> {
		// Get the module.
		let module = self.module_for_url(&params.text_document.uri).await?;

		// Get the document symbols.
		let symbols = self.symbols(&module).await?;
		let Some(symbols) = symbols else {
			return Ok(None);
		};

		// Convert the symbols.
		let symbols = symbols.into_iter().map(collect_symbol_tree).collect();

		Ok(Some(lsp::DocumentSymbolResponse::Nested(symbols)))
	}

	pub async fn symbols(&self, module: &Module) -> Result<Option<Vec<Symbol>>> {
		// Create the request.
		let request = super::Request::Symbols(Request {
			module: module.clone(),
		});

		// Perform the request.
		let response = self.request(request).await?;

		// Get the response.
		let super::Response::Symbols(response) = response else {
			return Err(error!("Unexpected response type."))
		};

		Ok(response.symbols)
	}
}

fn collect_symbol_tree(symbol: Symbol) -> lsp::DocumentSymbol {
	let Symbol {
		name,
		detail,
		kind,
		tags,
		range,
		selection_range,
		children,
	} = symbol;

	let kind = match kind {
		Kind::File => lsp::SymbolKind::FILE,
		Kind::Module => lsp::SymbolKind::MODULE,
		Kind::Namespace => lsp::SymbolKind::NAMESPACE,
		Kind::Package => lsp::SymbolKind::PACKAGE,
		Kind::Class => lsp::SymbolKind::CLASS,
		Kind::Method => lsp::SymbolKind::METHOD,
		Kind::Property => lsp::SymbolKind::PROPERTY,
		Kind::Field => lsp::SymbolKind::FIELD,
		Kind::Constructor => lsp::SymbolKind::CONSTRUCTOR,
		Kind::Enum => lsp::SymbolKind::ENUM,
		Kind::Interface => lsp::SymbolKind::INTERFACE,
		Kind::Function => lsp::SymbolKind::FUNCTION,
		Kind::Variable => lsp::SymbolKind::VARIABLE,
		Kind::Constant => lsp::SymbolKind::CONSTANT,
		Kind::String => lsp::SymbolKind::STRING,
		Kind::Number => lsp::SymbolKind::NUMBER,
		Kind::Boolean => lsp::SymbolKind::BOOLEAN,
		Kind::Array => lsp::SymbolKind::ARRAY,
		Kind::Object => lsp::SymbolKind::OBJECT,
		Kind::Key => lsp::SymbolKind::KEY,
		Kind::Null => lsp::SymbolKind::NULL,
		Kind::EnumMember => lsp::SymbolKind::ENUM_MEMBER,
		Kind::Event => lsp::SymbolKind::EVENT,
		Kind::Operator => lsp::SymbolKind::OPERATOR,
		Kind::TypeParameter => lsp::SymbolKind::TYPE_PARAMETER,
	};

	let tags = tags
		.into_iter()
		.map(|tag| match tag {
			Tag::Deprecated => lsp::SymbolTag::DEPRECATED,
		})
		.collect();

	let children = children.map(|children| children.into_iter().map(collect_symbol_tree).collect());

	let range = lsp::Range {
		start: lsp::Position {
			line: range.start.line,
			character: range.end.character,
		},
		end: lsp::Position {
			line: range.end.line,
			character: range.end.character,
		},
	};

	let selection_range = lsp::Range {
		start: lsp::Position {
			line: selection_range.start.line,
			character: selection_range.end.character,
		},
		end: lsp::Position {
			line: selection_range.end.line,
			character: selection_range.end.character,
		},
	};

	#[allow(deprecated)]
	lsp::DocumentSymbol {
		name,
		detail,
		kind,
		tags: Some(tags),
		range,
		selection_range,
		children,
		deprecated: None,
	}
}
